core/services/ocr2/plugins/s4/plugin.go

```diff
@@ -0,0 +1,288 @@
+		c.minQueryAddress = new(big.Int).Add(c.minQueryAddress, c.addressInteval)
+		c.maxQueryAddress = new(big.Int).Add(c.maxQueryAddress, c.addressInteval)
+		if c.maxQueryAddress.Cmp(s4.MaxAddress) > 0 {
+			c.maxQueryAddress = s4.MaxAddress
+		}
+		if c.minQueryAddress.Cmp(s4.MaxAddress) >= 0 {
+			c.minQueryAddress = s4.MinAddress
+			c.maxQueryAddress = new(big.Int).Add(c.minQueryAddress, c.addressInteval)
+		}
+	}
+}
+
+func (c *plugin) Query(ctx context.Context, _ types.ReportTimestamp) (types.Query, error) {
+	snapshot, err := c.orm.GetSnapshot(c.minQueryAddress, c.maxQueryAddress, pg.WithParentCtx(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(snapshot) == 0 {
+		return nil, nil
+	}
+
+	rows := make([]*Row, len(snapshot))
+	for i, row := range snapshot {
+		rows[i] = convertRow(row)
+	}
+	query, err := marshalRows(rows, c.minQueryAddress, c.maxQueryAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	c.advanceAddressRange()
+
+	return query, err
+}
+
+func (c *plugin) Observation(ctx context.Context, _ types.ReportTimestamp, query types.Query) (types.Observation, error) {
+	if err := c.orm.DeleteExpired(pg.WithParentCtx(ctx)); err != nil {
+		return nil, err
+	}
+
+	queryRows, minAddress, maxAddress, err := unmarshalRows(query)
+	if err != nil {
+		return nil, err
+	}
+
+	snapshot, err := c.orm.GetSnapshot(minAddress, maxAddress, pg.WithParentCtx(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	snapshotMap := make(map[key]*Row)
+	unconfirmedMap := make(map[key]*Row)
+	for _, row := range snapshot {
+		r := convertRow(row)
+		mkey := key{
+			address: r.Address,
+			slot:    uint(r.Slotid),
+		}
+		snapshotMap[mkey] = r
+		if !row.Confirmed {
+			unconfirmedMap[mkey] = r
+		}
+	}
+
+	observation := make([]*Row, 0)
+	for _, queryRow := range queryRows {
+		mkey := key{
+			address: queryRow.Address,
+			slot:    uint(queryRow.Slotid),
+		}
+		snapshotRow, ok := snapshotMap[mkey]
+		if ok && queryRow.Version < snapshotRow.Version {
+			observation = append(observation, queryRow)
+			delete(unconfirmedMap, mkey)
+		}
+	}
+
+	for _, unconfirmed := range unconfirmedMap {
+		observation = append(observation, unconfirmed)
+	}
+	if len(observation) > int(c.config.MaxObservationEntries) {
+		observation = observation[:c.config.MaxObservationEntries]
+	}
+
+	return marshalRows(observation, minAddress, maxAddress)
+}
+
+func (c *plugin) Report(_ context.Context, _ types.ReportTimestamp, _ types.Query, aos []types.AttributedObservation) (bool, types.Report, error) {
+	reportMap := make(map[key]*Row)
+
+	for _, ao := range aos {
+		observationRows, _, _, err := unmarshalRows(ao.Observation)
+		if err != nil {
+			return false, nil, err
+		}
+
+		for _, row := range observationRows {
+			mkey := key{
+				address: row.Address,
+				slot:    uint(row.Slotid),
+			}
+			report, ok := reportMap[mkey]
+			if ok && report.Version > row.Version {
+				continue
+			}
+			if err := verifySignature(row); err != nil {
+				return false, nil, err
```

Comment:
continue instead of return (+ log error as that most likely indicates a malicious node)